src/plugins/map/MQ2MapCommands.cpp

```diff
@@ -1169,151 +1182,350 @@ char* FormatMarker(const char* szLine, char* szDest, size_t BufferSize)
+	if (gOverrideMapLocParams.width != gDefaultMapLocParams.width) return true;
+	if (gOverrideMapLocParams.color.Red != gDefaultMapLocParams.color.Red) return true;
+	if (gOverrideMapLocParams.color.Green != gDefaultMapLocParams.color.Green) return true;
+	if (gOverrideMapLocParams.color.Blue != gDefaultMapLocParams.color.Blue) return true;
+	if (gOverrideMapLocParams.circleRadius != gDefaultMapLocParams.circleRadius) return true;
+	if (gOverrideMapLocParams.circleColor.Red != gDefaultMapLocParams.circleColor.Red) return true;
+	if (gOverrideMapLocParams.circleColor.Green != gDefaultMapLocParams.circleColor.Green) return true;
+	if (gOverrideMapLocParams.circleColor.Blue != gDefaultMapLocParams.circleColor.Blue) return true;
+
+	return false;
+}
+
+static void BuildFilteredOptionArray(std::vector<MapFilterOption*> &options)
+{
+	for (size_t childIndex = 0; childIndex < options.size(); childIndex++)
+	{
+		MapFilter parentFilter = options[childIndex]->RequiresOption;
+		if (parentFilter != MapFilter::Invalid)
+		{
+			MapFilterOption parentFilterOption = GetMapFilterOption(parentFilter);
+			for (size_t parentIndex = 0; parentIndex < options.size(); parentIndex++)
+			{
+				if (std::strcmp(parentFilterOption.szName, options[parentIndex]->szName) == 0)
+				{
+					// Find location to move child after parent, including sorting sub children alphabetically
+					int childNewIndex = parentIndex + 1;
+					for (size_t movedChildIndex = childNewIndex;
+						movedChildIndex < options.size() - 1 &&
+						options[movedChildIndex]->RequiresOption == options[movedChildIndex + 1]->RequiresOption &&
+						std::strcmp(options[movedChildIndex]->szName, options[movedChildIndex + 1]->szName) > 0;
+						movedChildIndex++)
+					{
+						childNewIndex = movedChildIndex;
+					}
+
+					move(options, childIndex, childNewIndex);
+
+					// Adjust iterator for movement
+					if (childIndex < parentIndex)
+					{
+						// Current element was moved later and the next element was put in its place, so recheck current.
+						childIndex--;
+					} // else current element was moved earlier in the array, and the element that took its place is one we previously checked, so continue
+
+					continue; // we found the parent, move onto next i value and skip the rest of the j's
+				}
+			}
+		}
+	}
+}
+
+static void BuildOptionArrays()
 {
+	// Build the lists once, cache them for future use
+	if (mapFilterOptions.size() > 0)
+		return;
+
+	// Copy MapFilterOptions
+	std::vector<MapFilterOption*> allFilterOptions;
+	for (MapFilterOption& option : MapFilterOptions)
+	{
+		if (option.ThisFilter != MapFilter::Invalid && option.ThisFilter != MapFilter::All)
+			allFilterOptions.push_back(&option);
+	}
+
+	// Sort copied array alphabetically
+	std::sort(allFilterOptions.begin(), allFilterOptions.end(),
+		[](MapFilterOption* a, MapFilterOption* b) -> bool
+		{
+			if (a == nullptr || a->szName == nullptr) return false;
+			if (b == nullptr || b->szName == nullptr) return true;
```

Comment:
in what cases can a/b be nullptr?

Reply:
this was before making the MapFilterOptions a vector I believe.